Element Templates Runtime Versions support

jarekdanielak opened the pull request:
Related to https://github.com/camunda/camunda-modeler/issues/4530
Related to https://github.com/camunda/product-hub/issues/2368 (epic)

### Proposed Changes

<!--

Add relevant context (issue fixed or related to), 
a capture of the UI changes (if any) as well as 
steps to try out your changes.

--> 


The goal of this pull request is to ensure that served element templates are compatible (best to our knowledge) with current runtime environment. 

Modelers will be able to pass their `engines` configuration to `elementTemplates`. If templates have information about compatible engines in their schema (introduced with https://github.com/camunda/element-templates-json-schema/issues/146), compatibility will be verified and relevant templates will be provided.

If runtime environment changes dynamically, it can be adjusted with `elementTemplates#setEngines` API.

Bunch of enhancements to support element templates runtime versions:
- [x] Add `engines` configuration to ElementTemplates module.
- [x] Add API to set `engines` config.
- [x] Add API to check if template is compatible with currently set engines.
- [x] Verify compatibility when setting templates and only serve compatible templates as latest.
- [x] Introduce new "Not compatible" state in properties panel element templates group to differentiate between not found and not compatible templates.
- [x] Add linting for incompatible template.
- [x] Clean up commits
- [ ] (optional) Console warning if templates provide compatibility info, but modeler didn't provide runtime version.

### Test locally

`npx @bpmn-io/sr bpmn-io/bpmn-js-element-templates#engines`

### Checklist

To ensure you provided everything we need to look at your PR:

* [x] **Brief textual description** of the changes present
* [ ] **Visual demo** attached
* [x] **Steps to try out** present, i.e. [using the `@bpmn-io/sr` tool](https://github.com/bpmn-io/sr)
* [x] Related issue linked via `Closes {LINK_TO_ISSUE}` or `Related to {LINK_TO_ISSUE}`

<!--

Thanks for creating this pull request! ❤️

-->

barmac commented:
Architecture sketch done in the session with @niku:

![image](https://github.com/user-attachments/assets/6f046ebf-b724-4a9e-bd6e-03b9be9c04dc)

jarekdanielak commented:
@nikku filled the tests and fixed small things on the way.

I moved `Validator` tests from `cloud-element-templates` to base as the engines validation happens in the base Validator.

nikku commented:
Turns out we did not add validation for `cloud*` templates, and also no test coverage.

Via https://github.com/bpmn-io/bpmn-js-element-templates/pull/132/commits/07b580e8a984f0240d9735086bca33c567bf65b0 I made sure that we have both in there, and tests are available for both `element-template` variants.

I find https://github.com/bpmn-io/bpmn-js-element-templates/pull/132/commits/d66afd3d61fa4b4dbeeba88079e34e834359da16 interesting @jarekdanielak. You add this to the manual test bed, but don't add it as an integration test?

What would be the expected behavior if _only_ incompatible templates are provided? We don't test cover this (yet).

jarekdanielak commented:
> I find https://github.com/bpmn-io/bpmn-js-element-templates/commit/d66afd3d61fa4b4dbeeba88079e34e834359da16 interesting @jarekdanielak. You add this to the manual test bed, but don't add it as an integration test?

It's covered [here](https://github.com/bpmn-io/bpmn-js-element-templates/pull/132/commits/fd05d685dc477ea40f1682b9fff5bb5739db4a2b#:~:text=should%20display%20incompatible%20button%20when%20template%20is%20incompatible). Or do you mean engine version zero?

> What would be the expected behavior if only incompatible templates are provided? We don't test cover this (yet).

First version (first provided in `templates` array) will be set as latest. `elementTemplates#set` doesn't validate compatibility if template is not yet in the `templatesById` dictionary. We later detect this as incompatibility in [ElementTemplatesGroup](https://github.com/bpmn-io/bpmn-js-element-templates/pull/132/files#diff-fd6bf402a5bd8ef26fcd426060dabfcacd19a05dc42e9bd87ae2544a1dbbb62c:~:text=const-,isNewerCompatible,-%3D%20elementTemplates.), but should we not set it in the first place?

nikku commented:
> First version (first provided in templates array) will be set as latest. elementTemplates#set doesn't validate compatibility if template is not yet in the templatesById dictionary. We later detect this as incompatibility in [ElementTemplatesGroup](https://github.com/bpmn-io/bpmn-js-element-templates/pull/132/files#diff-fd6bf402a5bd8ef26fcd426060dabfcacd19a05dc42e9bd87ae2544a1dbbb62c:~:text=const-,isNewerCompatible,-%3D%20elementTemplates.), but should we not set it in the first place?

Exactly, cf. https://github.com/bpmn-io/bpmn-js-element-templates/pull/132/commits/f37ae0efdbabce78f3dd79949c64534f23e87830.

nikku commented:
https://github.com/bpmn-io/bpmn-js-element-templates/pull/132/commits/2ab8f11a209ee573b1d10e4ea7803e8414fcac05 marks only compatible templates as latest.

nikku commented:
I suggest you @jarekdanielak and @lmbateman checkin and fine-tune the UI of the update buttons.

You can find my proposal [here](https://github.com/bpmn-io/bpmn-js-element-templates/pull/132/commits/229eee04fe35a64ef1f654c79e6033d594488171), which includes a "yellow", "warning" state that signals "you may better do something about this (but don't have to). I'm resistant to show incompatibility as :red_circle:, cf. [this conversation and following](https://camunda.slack.com/archives/GP70M0J6M/p1733775106831379?thread_ts=1733755501.655649&cid=GP70M0J6M):

![capture yqOAd3_optimized](https://github.com/user-attachments/assets/f4eab555-d092-41e0-9e22-b69829ccdd4b)

nikku commented on src/cloud-element-templates/linting/LinterPlugin.js:
```diff
@@ -37,12 +37,28 @@ export const elementTemplateLintRule = ({ templates = [] }) => {
   elementTemplates.set(validTemplates);
 
+  function setEngines(engines) {
+    if (engines == {}) {
```
`==` does not work on `{}`. `{} == {}` will always be false (different identity).

jarekdanielak replied:
Thanks. Let's wait with review until I push "full" linting after latest discussions.

barmac replied:
BTW I'd expect the linter to enforce `===`

nikku commented:
@jarekdanielak What do you think of the following: We add `elementTemplates` as a default provided `engine`, pinned to the `bpmn-js-element-templates` version.

Why? Because this allows environments to specify: "compatible with element templates >= xyz", so older Desktop Modeler versions would be able to safely ignore templates that are tagged for incompatible element template run-times.

nikku commented on src/element-templates/ElementTemplates.js:
```diff
@@ -196,24 +196,35 @@ export default class ElementTemplates {
+   *
+   * @return {Object} - incompatible engines along with their template and local versions
+   */
+  getIncompatibleEngines(template) {
```
If we expose another public API then we need to test this.

jarekdanielak replied:
Added with https://github.com/bpmn-io/bpmn-js-element-templates/pull/132/commits/54da72d8601a4af889f53738405aaee370165e9e

jarekdanielak commented:
Updated with linting and UI changes discussed in the [Slack thread](https://camunda.slack.com/archives/GP70M0J6M/p1733755501655649).

Since we want compatibility to lint with warning level, I extracted it to a new rule and reorganized linting code to make it easier to navigate and extend (https://github.com/bpmn-io/bpmn-js-element-templates/pull/132/commits/36cf7a506e56e6d9c1844c8bdb6c7671fb67c551).

---

I consider this now ready for review: @barmac, @philippfromme 

@lmbateman you can run it and check all the element templates states we now have with: 

`npx @bpmn-io/sr bpmn-io/bpmn-js-element-templates#engines`. 

Please give me your thumbs up or comments on the final design.

---

> What do you think of the following: We add elementTemplates as a default provided engine, pinned to the bpmn-js-element-templates version.

@nikku let's discuss tomorrow.

jarekdanielak commented:
Summary of discussion with @nikku about adding current version of `bpmn-js-element-templates` as a default engine:

Providing `elementTemplates` engine will allow template authors to make sure old modelers (using old version of `bpmn-js-element-templates`) don't load new, incompatible templates.

Let's build it. 👍

In the example below, it's clear that Template Version 2 relies on some new `bpmn-js-element-templates` features and shouldn't be used with older versions. 

![image](https://github.com/user-attachments/assets/1bac87c1-080e-4408-9d4d-4d0b43b19d92)

nikku commented:
This now looks good to me. Performed another minor cleanup, with minor changes:

* Do not leak `semver` internals via `elementTemplates#getEngines`, this is now a `Record<string, string>` object, and we have proper test coverage for it, too
* Additional tests to verify linting compatibility in the absence of `executionPlatform`, removed injection of `camundaDesktopModeler` through XML parsing (https://github.com/bpmn-io/bpmn-js-element-templates/pull/132/commits/ad8897ed97a108e1ee5094d9d6779a1db00574eb). Looking at last exporter is dangerous, as this field indicates the _last_, not necessarily the current editor.
* Simplified injection of `elementTemplates` version, preventing leak of `package.json` into bundled file (https://github.com/bpmn-io/bpmn-js-element-templates/pull/132/commits/6e04e560753d84e0c1a99338818527ca1c25c080), simplified extaction at lint time (https://github.com/bpmn-io/bpmn-js-element-templates/pull/132/commits/2c95fc5b5798d13d3e47def59ca2882f4892f4c6)

Please double-check, consider to squash https://github.com/bpmn-io/bpmn-js-element-templates/pull/132/commits/6e04e560753d84e0c1a99338818527ca1c25c080 and https://github.com/bpmn-io/bpmn-js-element-templates/pull/132/commits/ad8897ed97a108e1ee5094d9d6779a1db00574eb, and we're good to merge.

jarekdanielak commented:
The problem only happened in the Example test and is now fixed. Good catch. 👍

nikku commented:
The example test bed is broken now, after the fixes for https://github.com/bpmn-io/bpmn-js-element-templates/pull/132#issuecomment-2538398750:

![capture wheEcl_optimized](https://github.com/user-attachments/assets/b12471e2-625e-4394-a508-56bb514b141d)

We can follow-up with the fix after merging this PR, but my question would be: What is our stable pattern to hook up engine changes (in XML) _and_ `elementTemplates.setEngines()`? I don't see how/why we want to do it differently from this in the actual editors that implement a version chooser:

```javascript
const ChangeEnginesModule = {

  __init__: [ function(bpmnjs, eventBus, elementTemplates) {

    eventBus.on([
      'import.done',
      'elements.changed'
    ], function() {
      const executionPlatformVersion = bpmnjs.getDefinitions().get('executionPlatformVersion');

      elementTemplates.setEngines({
        camunda: executionPlatformVersion
      });
    });
  } ]
};
```

nikku commented:
Fixed "everything" at last, via https://github.com/bpmn-io/bpmn-js-element-templates/pull/132/commits/7f817669b91cc763ad70847b3fcd7c74504b4d26 and ensure the integration pattern works.

nikku reviewed (approved):
This works as expected, let's go forth and merge and release.
